### ISSUE TITLE
Some question about loss

### ISSUE THREAD

**Issue by lanslotttTT** [lzx551402/contextdesc#23]
In the paper you said that"In the training, we randomly sample a number of matchable keypoints as well as some undiscovered and unrepeatable keypoints (denoted as noisy keypoints), in order to have a complete simulation that is necessary to acquire strong generalization ability"
  But I see the loss were compute by matchable keypoint sets,I don't understand  how to use noisy keypoint during the training. What the effection about noisy keypoint?Are they need to compute loss separately?
  Thanks! : )

**Comment by vMagina** [comment]
这个的话，由于包含了noisy pair，会影响到正向传播的过程，好比是学习的特征中包含了噪声。然后在loss计算中只对可匹配的进行计算loss，不然肯定无法保证噪声点的内积大的。所以，你可以在训练的时候对每个image pair得到一个label，用来表示那些是真正的可匹配对，在计算loss时，根据label来对可匹配对计算loss。

**Comment by lanslotttTT** [comment]
主要是不懂原理，加入噪声的话好像N-pair loss 的矩阵 对角线上有噪声，就无法用softmax进一步计算loss了。N-pair loss 想让对角线上都是可匹配图像块之间的距离，加入label处理过之后，似乎又是在可匹配对上重新建立了一个矩阵。

**Comment by vMagina** [comment]
这个就是只对可匹配对构建的内积矩阵，所以Npair loss的矩阵里没有噪声对。loss的计算里面没有噪声点，可以看附录里面的公式，loss是在可匹配对上做的计算。加入噪声其实是想模拟真实情况下有噪声的情况，通过CN便可以学习到分布。最终，拟合的还是可匹配点对。

**Comment by lanslotttTT** [comment]
这样的话不加噪声好像对训练网络没什么影响，更新网络的loss都是可匹配对，没有噪声对。总觉得应该对噪声对进行某些处理来辅助训练网络

**Comment by lzx551402** [comment]
Sorry for the late reply since I was busy with graduation. Basically the understanding of @lanslotttTT is correct, the terminology "noisy keypoints" might be a bit confusing, which should have been more properly referred to as "unmatchable keypoints".

The key insight to introduce "unmachable keypoints" during training is to simulate the "real" matching process as much as possible. During inference, you cannot assume all keypoints are matchable (since the keypoint detector is imperfect, and only part of image is co-visible). 

And yes, the loss is only computed from "matchable keypoints". However, since we take row-(column-) wise softmax on the entire distance matrix, even "unmatchable keypoints" would influence the loss computation. Meanwhile, the "unmatchable keypoints" are necessary for training the geometric context. Again, you cannot assume that every keypoint is matchable during testing, and such "noise" must be considered during training.

**Comment by lzx551402** [comment]
@vMagina @lanslotttTT Hi, we have finally resolved the license issue, and made the training code publicly available. Please visit [TFMatch](https://github.com/lzx551402/tfmatch) if you are interested.

**Comment by lanslotttTT** [comment]
Thank you for your contribution, it will be very helpful for our learners。
